Improve type label support

jrgissing opened the pull request:
**Summary**

add direct type label support for commands/fixes/computes that have 'type' arguments

- [x] create_atoms command
- [x] fix deposit
- [x] fix mol/swap
- [x] fix atom/swap
- [x] group command
- [ ] fix shake/rattle
  - [x] implemented but not committed. type labels could pretty easily interfere with 'a', 'b', etc. keywords
- [x] fix gcmc
- [x] fix widom
- [x] delete_bonds command
- [x] fix bond/create/break
- [x] fix adapt + adapt/fep
  - [x] add option to feed a type label into utils::bounds. returns nlo = nhi = label2type(type_label)
  - [x] I < J requirement relaxed for non-asterisk syntax
- [x] compute rdf
  - [x] could technically break there was a type label called 'cutoff'
- [ ] pair_style tracker
  - [x] not sure if can work, because commas used in lists
- [x] fix charge/regulation
- [ ] variable command (type==H)
  - [x] rely on label2type(kind,label)
- [x] Dump Output Support
  - [x] dump_modify 'types' for xyz
  - [x] 'typelabel' keyword for dump custom
- [ ] compute property/local - will require refactor of communication routines, suggest a separate PR
- [ ] others?

Existing/prior support includes:
- [x] data file/molecule templates
- [x] label2type function
- [x] set command
- [x] coeff commands
- [x] mass command

**Related Issue(s)**

none

**Author(s)**

Jake

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

yes

**Implementation Notes**

new string output (type labels) to dump files not added in a very extensible manner. can refactor if you think more string fields will be added to dump output

~~currently, 'dump_modify types numeric' (only relevant to dump xyz) does not revert a previous 'dump_modify types labels' command~~

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

jrgissing committed:
create_atoms: direct type label support

jrgissing committed:
fix_deposit: direct type label support

jrgissing committed:
fix/mol/swap: direct type label support

jrgissing committed:
refactor expand_types to return int

jrgissing committed:
make use of refactored expand_type elsewhere

jrgissing committed:
cleanup

jrgissing committed:
generalize passthrough args

jrgissing committed:
type label support for atom/swap

jrgissing committed:
direct type label for group

jrgissing committed:
remove bogus example

akohlmey commented on doc/src/create_atoms.rst:
```diff
@@ -74,7 +74,13 @@ Examples
 .. code-block:: LAMMPS
 
    create_atoms 1 box
-   create_atoms 3 region regsphere basis 2 3
```
Please don't delete examples for numerical inputs

jrgissing committed:
restore numerical examples

(duplicated in type label versions)

jrgissing committed:
restore numerical types example

(duplicated with type labels)

akohlmey committed:
make pip install packages in virtual environment

akohlmey committed:
downgrade macOS to version 13

jrgissing committed:
delete_bonds: direct type label support

jrgissing committed:
fix_gcmc: direct type label support

apparently changed text in doc description is just reflowing

jrgissing committed:
fix_widom: direct type label support

jrgissing committed:
bond/create: direct type label support

jrgissing committed:
start off with simpler labelmap example

jrgissing committed:
bond/break: direct type label support

jrgissing committed:
fix adapt: direct type label support

make utils::bound type aware?
i.e., Atom:BOND argument instead of atom->nbondtypes

jrgissing committed:
adapt/fep: direct type label support

jrgissing committed:
finish adapt docs

jrgissing committed:
typos

jrgissing committed:
example with augmented utils::bounds

option to check for type label

jrgissing committed:
Revert "example with augmented utils::bounds"

This reverts commit 25d4b3484d83913e3825e7b9d0ef013fa02a461b.

jrgissing committed:
bounds() wrapper for type labels

jrgissing committed:
fix/adapt/fep: use bounds_typelabel

jrgissing committed:
compute_rdf: direct type label support

type label that is the same as the keyword ('cutoff') will break things. if syntax is otherwise 'correct', then will throw a syntax error. perhaps could run through typelabels to check first?

jrgissing committed:
more uses of bounds_typelabel

jrgissing committed:
charge/regulation: direct type label support

jrgissing committed:
add citeme

jrgissing committed:
Merge branch 'lammps:develop' into improve_type_label_support

stanmoore1 commented:
Still to do: add support for dump files

jrgissing committed:
type label support for dump xyz

jrgissing committed:
make use of new dump_modify in examples

jrgissing committed:
dump custom, 'typelabel' attribute

writing strings to dump files previously was not implemented in general way.
did not refactor to make more general.
NOTE: added value to middle of enum

jrgissing commented:
> Still to do: add support for dump files

dump file support is complete

jrgissing committed:
let dump_modify types numeric revert labels

jrgissing committed:
Merge branch 'lammps:develop' into improve_type_label_support

akohlmey committed:
Merge branch 'develop' into improve_type_label_support

akohlmey commented:
@sjplimp @rbberger @athomps @jtclemm This is now flagged as ready for review and needs to be reviewed.

akohlmey commented:
@jrgissing It would be great if there would be some tests in the unittest tree added for the new functionality.

jrgissing commented:
> @jrgissing It would be great if there would be some tests in the unittest tree added for the new functionality.

Good idea. I am in a bit of a crunch for the next week and a half, but I will add some unit tests after that

akohlmey commented:
> > @jrgissing It would be great if there would be some tests in the unittest tree added for the new functionality.
> 
> Good idea. I am in a bit of a crunch for the next week and a half, but I will add some unit tests after that

Ok. Then I'll merge this "as is" now, so it will be in the next feature release and the tests can be in a new pull requests. I may be able to do a few of them on the side (the simple ones)
